### PR TITLE
Seek intron, 3'-UTR, and 5'-UTR regions from genomic position

### PR DESCRIPTION
**Pull request by nuggetoriniku**
The seek-gene-region function can return Intron and UTR.

I'm trying to split the key `:exon-foo` into `:type` and `:foo` .
 - The value of `:type` is a region type such as 'exon',  'intron', etc ...
 - The value of `:foo` is conventional (index, count, etc.)

### CODE REVIEW

File: src/varity/ref_gene.clj [hunk 1]
```diff
@@ -213,22 +213,49 @@
...
+                                       txe (:tx-end rg)]
+                                   (cond
+                                     (< pos txs) (first (sgn '({:region "UTR-5" :idx 0 :count 1}
+                                                               {:region "UTR-3" :idx 0 :count 1})))
+                                     (> pos txe) (second (sgn '({:region "UTR-5" :idx 0 :count 1}
+                                                                {:region "UTR-3" :idx 0 :count 1})))
+                                     :else (if exon-idx
+                                             {:region "exon" :idx exon-idx :count (count exon-ranges)}
+                                             {:region "intron" :idx intron-idx :count (count intron-ranges)})))]
+
```
Comment (federkasten) [inline src/varity/ref_gene.clj:253]:
```suggestion
```

File: src/varity/ref_gene.clj [hunk 2]
```diff
@@ -213,22 +213,49 @@
...
+                                            ((fn [r]
+                                               (loop [intron-ranges []
+                                                      left (first r)
+                                                      right (second r)
+                                                      ranges (rest r)]
+                                                 (if (empty? (first ranges)) intron-ranges
+                                                     (recur (conj intron-ranges [(inc (second left)) (dec (first right))])
+                                                            right
+                                                            (second ranges)
+                                                            (rest ranges))))))
```
Comment (federkasten) [inline src/varity/ref_gene.clj:238]: Could you define as an independent function such as `exon-ranges->intron-ragnges`? Additionally, please add relevant testing codes for it.

File: src/varity/ref_gene.clj [hunk 3]
```diff
@@ -213,22 +213,49 @@
...
+                                            (sgn)))
+                     intron-idx (->> intron-ranges
+                                     (keep-indexed (fn [i [s e]] (if (<= s pos e) i)))
+                                     first)
+                     region-type (let [txs (:tx-start rg)
+                                       txe (:tx-end rg)]
```
Comment (federkasten) [inline src/varity/ref_gene.clj:244]: These `txs` and `txe` bindings are redundant.

File: src/varity/ref_gene.clj [hunk 4]
```diff
@@ -213,22 +213,49 @@
...
+                                     :else (if exon-idx
+                                             {:region "exon" :idx exon-idx :count (count exon-ranges)}
+                                             {:region "intron" :idx intron-idx :count (count intron-ranges)})))]
+
+                 {:region (:region region-type) 
+                  :index (if (:idx region-type)
+                           (inc (:idx region-type))
+                           nil)
+                  :count (:count region-type)
                   :gene rg}))))))
```
Comment (federkasten) [inline src/varity/ref_gene.clj:259]: I think you can write more simply with `update` function as follows:

```clj
(update region-type :index inc)
```

File: src/varity/ref_gene.clj [hunk 5]
```diff
@@ -213,22 +213,49 @@
...
+                     region-type (let [txs (:tx-start rg)
+                                       txe (:tx-end rg)]
+                                   (cond
+                                     (< pos txs) (first (sgn '({:region "UTR-5" :idx 0 :count 1}
+                                                               {:region "UTR-3" :idx 0 :count 1})))
+                                     (> pos txe) (second (sgn '({:region "UTR-5" :idx 0 :count 1}
+                                                                {:region "UTR-3" :idx 0 :count 1})))
+                                     :else (if exon-idx
+                                             {:region "exon" :idx exon-idx :count (count exon-ranges)}
+                                             {:region "intron" :idx intron-idx :count (count intron-ranges)})))]
```
Comment (federkasten) [inline src/varity/ref_gene.clj:252]:
```suggestion
                                     exon-idx
                                     {:region "exon" :idx exon-idx :count (count exon-ranges)}
                                     
                                     intron-idx
                                     {:region "intron" :idx intron-idx :count (count intron-ranges)}
                                     
                                     :else
                                     nil)))]
```

`if` is nested in `cond` here. I prefer the flattened description, such as suggested above.

**Comment by nuggetoriniku** [comment]
Thank you for your review.
I have improved my code based on your review.
Could you check again ? :pray:

File: src/varity/ref_gene.clj [hunk 6]
```diff
@@ -208,28 +208,56 @@
...
-                  :exon-count (count exon-ranges)
-                  :gene rg}))))))
+               (let [sgn (case (:strand rg)
+                           :forward identity
+                           :reverse reverse)
+                     exon-ranges (:exon-ranges rg)
+                     exon-idx (->> (sgn exon-ranges)
+                                   (keep-indexed (fn [i [s e]] (if (<= s pos e) i)))
+                                   first)
+                     intron-ranges (if exon-idx nil
```
Comment (federkasten) [inline src/varity/ref_gene.clj:241]:
```suggestion
                     intron-ranges (if-not exon-idx
```

File: src/varity/ref_gene.clj [hunk 7]
```diff
@@ -208,28 +208,56 @@
...
+                                   (> pos (:tx-end rg)) (second (sgn '({:region "UTR-5" :index 0 :count 1}
+                                                                       {:region "UTR-3" :index 0 :count 1})))
+                                   exon-idx {:region "exon" :index exon-idx :count (count exon-ranges)}
+                                   intron-idx {:region "intron" :index intron-idx :count (count intron-ranges)}
+                                   :else nil)]
+                 (-> region-type
+                     (update :index (fn [idx]
+                                      (if idx (inc idx)
+                                          nil)))
+                     (merge {:gene rg}))))))))
```
Comment (federkasten) [inline src/varity/ref_gene.clj:260]:
```suggestion
                     (assoc :gene rg))))))))
```

File: src/varity/ref_gene.clj [hunk 8]
```diff
@@ -208,28 +208,56 @@
...
+                                                                        {:region "UTR-3" :index 0 :count 1})))
+                                   (> pos (:tx-end rg)) (second (sgn '({:region "UTR-5" :index 0 :count 1}
+                                                                       {:region "UTR-3" :index 0 :count 1})))
+                                   exon-idx {:region "exon" :index exon-idx :count (count exon-ranges)}
+                                   intron-idx {:region "intron" :index intron-idx :count (count intron-ranges)}
+                                   :else nil)]
+                 (-> region-type
+                     (update :index (fn [idx]
+                                      (if idx (inc idx)
+                                          nil)))
```
Comment (federkasten) [inline src/varity/ref_gene.clj:259]:
```suggestion
                                      (when idx (inc idx))))
```

File: src/varity/ref_gene.clj [hunk 9]
```diff
@@ -208,28 +208,55 @@
...
+       ((fn [r]
+          (loop [intron-ranges []
+                 left (first r)
+                 right (second r)
+                 ranges (rest r)]
+            (if (empty? (first ranges)) intron-ranges
+                (recur (conj intron-ranges [(inc (second left)) (dec (first right))])
+                       right
+                       (second ranges)
+                       (rest ranges))))))))
```
Comment (federkasten) [inline src/varity/ref_gene.clj:223]: I think this arrow macro `->>` is not elegant. It's better to describe in simple `(loop [] ... (recure)` way.

Comment (nuggetoriniku) [reply src/varity/ref_gene.clj:223]: I'm sorry, you are right. 🙇‍♂️

Comment (athos) [reply src/varity/ref_gene.clj]: I’m not a designated reviewer, but let me add a few comments on this.

You can rewrite the code more simply using [destructing](https://clojure.org/guides/destructuring):

```clojure
(defn exon-ranges->intron-ranges
  [exon-ranges]
  (loop [intron-ranges []
         [left & [right :as ranges]] exon-ranges]
    (if (empty? right)
      intron-ranges
      (recur (conj intron-ranges [(inc (second left)) (dec (first right))])
             ranges))))
```

Or if you don’t mind additional allocation here, `partiton` will make it even shorter:

```clojure
(defn exon-ranges->intron-ranges
  [exon-ranges]
  (mapv (fn [[left right]] [(inc (second left)) (dec (first right))])
        (partition 2 1 exon-ranges)))
```

The more code you write, the more likely it is to have a bug in it, so it’s good to use existing powerful features effectively and keep our code simple as far as we can.

Comment (nuggetoriniku) [reply src/varity/ref_gene.clj]: Thank you for your review.

I have learned about `partition` function.
The code using `partition` looks more elegant than old one.

I'd like to write code with simplicity in mind ...

**Comment by nuggetoriniku** [comment]
@federkasten 
I changed condition of UTR. (:tx-start/:tx-end → :cds-start/:cds-end)
And format of the return value was changed.
Could you review this?

File: src/varity/ref_gene.clj [hunk 10]
```diff
@@ -208,28 +208,49 @@
...
+                                                                {:region "UTR-3" :index 0 :count 1})))
+                           :else nil)
+                     exon-intron (cond
+                                   exon-idx {:region "exon" :index exon-idx :count (count exon-ranges)}
+                                   intron-idx {:region "intron" :index intron-idx :count (count intron-ranges)})
+                     region-types (remove nil? (vector utr exon-intron))]
+                 (update {:region-types region-types :gene rg}
+                         :region-types
+                         (partial map #(update % :index (fn [idx]
+                                                          (when idx (inc idx))))))))))))
```
Comment (federkasten) [inline src/varity/ref_gene.clj:253]: I think it's a bit tricky description. I want to keep it simple here, as follows.


```suggestion
                 {:regions regions :gene rg}))))))
```

Then, I think it's easier to understand if you apply `inc` to `index` in the `exon-intron` at line [246 to 248](https://github.com/chrovis/varity/pull/36/files#diff-63a0f54cbd3574ddea6aa13e18ee93d0R246-R248).

```clj
                     exon-intron (cond
                                   exon-idx {:region "exon" :index (inc exon-idx) :count (count exon-ranges)}
                                   intron-idx {:region "intron" :index (inc intron-idx) :count (count intron-ranges)})
```

Besides, sorry to be so detailed, but I prefer `regions` to `region-types`.

**Review by federkasten — COMMENTED** [review]
Thanks. The changes are reasonable. I commented on some trivial points.

File: src/varity/ref_gene.clj (hunk 10 above)
Comment (nuggetoriniku) [reply src/varity/ref_gene.clj]: Thank you for your review :pray:
I have reflected your comment.

**Review by federkasten — APPROVED** [review]
LGTM :ok_man:

**Comment by federkasten** [comment]
@totakke  Sorry. Could you give a quick review before merging? 🙏

File: test/varity/ref_gene_test.clj [hunk 11]
```diff
@@ -191,16 +191,31 @@
...
+                                (mapv (fn [rt]
+                                        (mapv #(vector (:region %) (:index %) (:count %))
+                                              rt)))))
+        "chr4" 54736520 nil [[["exon" 18 21]] [["exon" 18 21]]]
+        "chr7" 116771976 "NM_000245" [[["exon" 14 21]]]
+        "chrX" 61197987 nil []
+        "chr3" 41224090 "NM_001904" [[["intron" 2 14]]]
+        "chr5" 12575053 nil [[["UTR-5" nil nil] ["intron" 1 3]]]
+        "chr10" 79512600 "NM_001099692" [[["UTR-5" nil nil]]]
+        "chr12" 101128642 "NM_001286615" [[["UTR-3" nil nil]] [["UTR-5" nil nil]]]
```
Comment (totakke) [inline test/varity/ref_gene_test.clj:217]: These tests do not include exon/intron regions of a reverse-strand gene. I recommend adding such a test case.

e.g. BRAF (strand: -)

```clj
"chr7" 140753336 "NM_004333" [[["exon" 15 18]]]
```

**Review by totakke — CHANGES_REQUESTED** [review]
Thank you for PR. I added a few comments.

**Review by totakke — APPROVED** [review]
LGTM

**Comment by totakke** [comment]
@federkasten I've bumped up development version to `0.7.0-SNAPSHOT`. You can merge this PR whenever you want.

**Comment by federkasten** [comment]
Thank you @totakke. I'll merge.

**Pull request merged**